extensions/security/runtime-spi/src/main/java/io/quarkus/security/spi/runtime/SecurityEventHelper.java

```diff
@@ -0,0 +1,93 @@
+package io.quarkus.security.spi.runtime;
+
+import java.util.Map;
+
+import jakarta.enterprise.event.Event;
+import jakarta.enterprise.inject.spi.BeanManager;
+
+import io.quarkus.security.identity.SecurityIdentity;
+
+public class SecurityEventHelper<S extends SecurityEvent, F extends SecurityEvent> {
+
+    public static final AuthorizationFailureEvent AUTHORIZATION_FAILURE = new AuthorizationFailureEvent(null, null, null);
+    public static final AuthorizationSuccessEvent AUTHORIZATION_SUCCESS = new AuthorizationSuccessEvent(null, null);
+    public static final AuthenticationFailureEvent AUTHENTICATION_FAILURE = new AuthenticationFailureEvent(null, null);
+    public static final AuthenticationSuccessEvent AUTHENTICATION_SUCCESS = new AuthenticationSuccessEvent(null, null);
+    private final boolean fireEventOnSuccess;
+    private final boolean fireEventOnFailure;
+    private final Event<S> successEvent;
+    private final Event<F> failureEvent;
+
+    public SecurityEventHelper(Event<S> successEvent, Event<F> failureEvent, S successInstance, F failureInstance,
+            BeanManager beanManager, boolean enabled) {
+        if (enabled) {
+            boolean fireAllSecurityEvents = !beanManager.resolveObserverMethods(new SecurityEvent() {
```

Comment:
Did you do any measurements? I'm still wondering if this optimization makes sense. I mean if no observers are present the `EventImpl#fire()` should be basically a no-op so :shrug: .

Reply:
> Did you do any measurements? I'm still wondering if this optimization makes sense. I mean if no observers are present the EventImpl#fire() should be basically a no-op so

No, 

- one, because such addition won't have effect as you correctly assume, but if we all write code that do things that are not necessary and keep adding features executed on every single request while users are waiting and not using these features, it will have effect.
- two, I'm not worried about `EventImpl#fire()`; why should I create maps, functions, add try-catch during every single HTTP request when most applications might not use it? Having this behind boolean limits cost of this PR. 

Let me you few examples:

1.  this handler https://github.com/quarkusio/quarkus/blob/008451f0c1ae3e80c2e6d093fd3cbe44feb85f28/extensions/resteasy-reactive/quarkus-resteasy-reactive/runtime/src/main/java/io/quarkus/resteasy/reactive/server/runtime/security/EagerSecurityHandler.java is created for every RR runtime resource method and I will be running runnable for success and throwable for failure even though I know I don't need it, then create map, unwrap `RoutingContext`, get check name and create event instance
2. I'll catch exception https://github.com/quarkusio/quarkus/blob/008451f0c1ae3e80c2e6d093fd3cbe44feb85f28/extensions/security/runtime/src/main/java/io/quarkus/security/runtime/interceptor/SecurityConstrainer.java#L63 which IIUC will only cost when actually catching it or create these mappers without needing it https://github.com/quarkusio/quarkus/blob/008451f0c1ae3e80c2e6d093fd3cbe44feb85f28/extensions/security/runtime/src/main/java/io/quarkus/security/runtime/interceptor/SecurityConstrainer.java#L82

I just don't think it's good idea to create things during requests I know ahead noone is using once in constructor of singleton / application scoped bean.

I'd agree with you if code was like `event.fire(new Event());`. Anyway, if it turns out your view is generally shared, I can yield.

Reply:
Ok, I do understand your points and I share your opinion that we should avoid unnecessary logic if possible. So let's keep it this way for now...